xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+) -> st.SearchStrategy[xr.Variable]:
+    """
+    Generates arbitrary xarray.Variable objects.
+
+    Follows the signature of the xarray.Variable constructor, but you can also pass alternative strategies to generate
+    either numpy-like array data or dimension names.
+
+    Passing nothing will generate a completely arbitrary Variable (backed by a numpy array).
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    data: Strategy generating array-likes, optional
+        Default is to generate numpy data of arbitrary shape, values and dtype.
+    dims: Strategy for generating the dimensions, optional
+        Can either be a strategy for generating a list of string dimension names,
+        or a strategy for generating a mapping of string dimension names to integer lengths along each dimension.
+        If provided in the former form the lengths of the returned Variable will either be determined from the
+        data argument if given or arbitrarily generated if not.
+        Default is to generate arbitrary dimension names for each axis in data.
+    attrs: Strategy which generates dicts, optional
+    convert: Callable
+        Function which accepts one numpy array and returns one numpy-like array of the same shape.
+        Applied to the data after it is drawn from the `data` strategy provided.
+        Useful for converting numpy arrays to other types of arrays, e.g. sparse arrays.
+        Default is a no-op.
+    """
+
+    if any(
+        not isinstance(arg, st.SearchStrategy) and arg is not None
+        for arg in [data, dims, attrs]
+    ):
+        raise TypeError(
+            "Contents must be provided as a hypothesis.strategies.SearchStrategy object (or None)."
+            "To specify fixed contents, use hypothesis.strategies.just()."
+        )
+
+    if data is not None and dims is None:
+        # no dims -> generate dims to match data
+        data = draw(data)
+        dims = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+
+    elif dims is not None and data is None:
+        # no data -> generate data to match dims
+        dims = draw(dims)
+        if isinstance(dims, List):
+            valid_shapes = npst.array_shapes(min_dims=len(dims), max_dims=len(dims))
+            data = draw(np_arrays(shape=draw(valid_shapes)))
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            shape = tuple(dims.values())
+            data = draw(np_arrays(shape=shape))
+
+    elif data is not None and dims is not None:
+        # both data and dims provided -> check drawn examples are compatible
+        dims = draw(dims)
+
+        # TODO is there another way to enforce these assumptions? This is very like to fail hypothesis' health checks
+        # TODO how do I write a test that checks that the hypothesis Unsatisfiable error will be raised?
+        # TODO or we could just raise in this case?
```

Comment:
I'd make it an explicit error to provide non-matching data and dims, because the performance cliff is basically vertical here.  

(for the test, `with pytest.raises(Unsatisfiable):`?  it's just a normal exception)

Reply:
If we're happy to raise explicit errors then can't we just use that approach to all combinations of kwargs? And just raise in all cases that would have terrible performance?

Reply:
I'm now raising an explicit `Unsatisfiable` error in the case of non-matching data and dims.